### ISSUE TITLE
snapshot() very slow

### ISSUE THREAD

**Issue by krlmlr** [rstudio/packrat#224]
In a project with 80 packages, `snapshot()` takes roughly two minutes on my machine. I think the reason is the recursive call to `getPackageRecords()` which seems to walk *all* dependency trees for *each* dependency. This means enumeration of all possible paths, and is troublesome e.g. for diamond structures:

A -> B -> D
A -> C -> D
D -> E -> G
D -> F -> G
G -> H -> J
G -> I -> J
...

Perhaps a caching environment can help: this env holds all package records that have been computed so far, should be consulted before doing any computations, and populated with already computed package records. However, the vectorized nature of `getPackageRecords` complicates things, a mapping from `pkgNames` to `allRecords` will be required.

**Comment by kevinushey** [comment]
I agree this is a good idea. This is an operation where vectorization really doesn't give us huge benefits, and so I agree just using a for loop + a caching environment would be a better solution.

**Comment by krlmlr** [comment]
Thanks. I'll look into it.

**Comment by IrfyNaz** [comment]
+1 - also seeing very slow snapshots on a project with ~ 100 packages, takes 20 mins to snapshot.
